ExPlat: Self-Contained Client Internals (requests only)

jessie-ross opened the pull request:
**This PR adds internal functions for the Explat Self-contained client.**

- Only `request.ts` in this one.

Part of a series rebuilding the reference WIP PR #48475


#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* See added Automated Tests 
* `yarn test-packages explat-client`


<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

jessie-ross committed:
Add ExPlat SAC package

jessie-ross committed:
Implement and test internals

jessie-ross committed:
Improve type imports

jessie-ross committed:
Update types

jessie-ross committed:
Document state functions

jessie-ross committed:
Minor name fix

jessie-ross committed:
Spell EA out

jessie-ross committed:
Use devDeps for calypso-polyfills

jessie-ross committed:
Test requests

jessie-ross committed:
Minor fix

yanirs commented on packages/explat-client/src/internal/requests.ts:
```diff
@@ -0,0 +1,88 @@
+	);
+
+	const fetchedExperimentAssignments = Object.entries( variations )
+		.map( ( [ experimentName, variationName ] ) => ( {
+			experimentName,
+			variationName,
+			retrievedTimestamp,
+			ttl,
+		} ) )
+		.map( validateExperimentAssignment );
```
Not critical at all given the expected size of the array, but how about we combine the functions to avoid iterating over the entries twice?

```javascript
.map( ( [ experimentName, variationName ] ) => validateExperimentAssignment( {
			experimentName,
			variationName,
			retrievedTimestamp,
			ttl,
		} ) )
```

jessie-ross replied:
I prefer the two maps, they separate two distinct operations.

yanirs commented on packages/explat-client/src/internal/test/requests.ts:
```diff
@@ -0,0 +1,233 @@
+const mockedConfig: Config = {
+	fetchExperimentAssignment: mockedFetchExperimentAssignment,
+	getAnonId: mockedGetAnonId,
+};
```
Since mocking / overriding the other two items is trivial, I think we might as well do it rather than disable the warnings. Something like:

```suggestion
const mockedConfig: Config = {
	fetchExperimentAssignment: mockedFetchExperimentAssignment,
	getAnonId: mockedGetAnonId,
	logError: () => {},
	isDevelopmentMode: false,
};
```

With this, we're also more explicit about not testing development mode.

yanirs commented on packages/explat-client/src/internal/test-common.ts:
```diff
@@ -12,3 +12,6 @@ export const validFallbackExperimentAssignment = {
 	isFallbackExperimentAssignment: true,
 };
+
+export const delayedValue = < T >( value: T, delayMilliseconds: number ): Promise< T > =>
```
I see that delayMilliseconds is always passed in as 1 in the new tests. Maybe make it optional and default to 1 so we can simply call `delayedValue( value )`?

jessie-ross replied:
I think it is better to be explicit here, also for some tests I use `0`.

yanirs replied:
Hmm... It's a bit hard to see what the 1 is referring to when reading the calling code. Maybe create constants and pass them in explicitly then? Something like this:

```javascript
const NO_DELAY = 0;
const MIN_DELAY = 1;

delayedValue( <value>, MIN_DELAY );
```

jessie-ross replied:
I'm using `ZERO_DELAY` and `ONE_DELAY` as I think `NO` and `MIN` are misleading:
- `NO_DELAY` won't be run instantly, instead it will be behaviourally equivalent to `setTimeout(f, 0)`
- `MIN_DELAY`: 0 is the min delay not 1.

jessie-ross replied:
Also I am not using it in `timing.ts` as many other functions are using millisecond time values and these values need to be easily seen and compared - especially with mock timers now being used.

yanirs commented on packages/explat-client/src/internal/test/requests.ts:
```diff
@@ -0,0 +1,233 @@
+			],
+		] );
+
+		mockedFetchExperimentAssignment.mockReset();
+		mockedGetAnonId.mockImplementationOnce( () => delayedValue( null, 1 ) );
+		mockFetchExperimentAssignmentToMatchExperimentAssignment( validExperimentAssignment );
+		const experimentAssignmentWithoutAnonId = await Requests.fetchExperimentAssignment(
+			mockedConfig,
+			validExperimentAssignment.experimentName
+		);
+
+		expect( experimentAssignmentWithoutAnonId ).toEqual( validExperimentAssignment );
+		expect( mockedFetchExperimentAssignment.mock.calls ).toEqual( [
+			[
+				{
+					anonId: null,
+					experimentName: 'experiment_name_a',
+				},
+			],
+		] );
```
It'd be nice to split this into a separate `it()` scenario since it's pretty much independent from the first scenario. That is, have `it( 'should successfully fetch and return a well formed response with an anon ID' )` and `it( 'should successfully fetch and return a well formed response without an anon ID' )`.

yanirs commented on packages/explat-client/src/internal/test/requests.ts:
```diff
@@ -0,0 +1,233 @@
+	it( 'should return true for valid reponses', () => {
+		expect(
+			Requests.isFetchExperimentAssignmentResponse( {
+				ttl: 0,
```
I wonder if this should be a valid response or if we should require `ttl` to be positive...

jessie-ross replied:
Yeah good point, if the `ttl` is positive that should safe guard us against too many requests being sent.

yanirs reviewed (approved):
Looks good! See some minor comments inline.

jessie-ross committed:
Merge branch 'trunk' into add/explat-sac-internals-request

jessie-ross committed:
Use Promise.resolve

jessie-ross committed:
Improve requests and related tests

- Ensure ttl is greater than 0
- Use constansts for 0 and 1 delays
- Minor cleanup

yanirs commented on packages/explat-client/src/internal/test/requests.ts:
```diff
@@ -0,0 +1,255 @@
+} );
+
+describe( 'isFetchExperimentAssignmentResponse', () => {
+	it( 'should return true for valid reponses', () => {
```
```suggestion
	it( 'should return true for valid responses', () => {
```

yanirs commented on packages/explat-client/src/internal/test/requests.ts:
```diff
@@ -0,0 +1,255 @@
+		).toBe( true );
+	} );
+
+	it( 'should return false for reponses with 0 ttl', () => {
```
```suggestion
	it( 'should return false for responses with 0 ttl', () => {
```

jessie-ross committed:
Update packages/explat-client/src/internal/test/requests.ts

Co-authored-by: Yanir Seroussi <yanir.seroussi@automattic.com>

jessie-ross committed:
Update packages/explat-client/src/internal/test/requests.ts

Co-authored-by: Yanir Seroussi <yanir.seroussi@automattic.com>
